File: accepted/235-lowest-common-ancestor-of-a-binary-search-tree.py
# https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {TreeNode} root
    # @param {TreeNode} p
    # @param {TreeNode} q
    # @return {TreeNode}
    def getPath(self, root, node):
        cur = root
        ret = []
        while cur.val != node.val:
            ret.append(cur)
            cur = cur.left if node.val < cur.val else cur.right
        ret.append(cur)
        return ret
    
    def lowestCommonAncestor(self, root, p, q):
        if None == root:
            return None
        p_path = self.getPath(root, p)
        q_path = self.getPath(root, q)
        
        ret = root
        i = 1
        
        while i < len(p_path) and i < len(q_path) and p_path[i] == q_path[i]:
            ret = p_path[i]
            i += 1
        
        return ret

File: accepted/235-lowest-common-ancestor-of-a-binary-search-tree.c
/* https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/ */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) 
{
    struct TreeNode *pCur = root;
    struct TreeNode *pRet = pCur;
    
    if (p == q)
        return p;
    
    while (NULL != pCur)
    {
        pRet = pCur;
        if (pCur == p || pCur == q)
            break;
        
        struct TreeNode *pNextP = p->val < pCur->val ? pCur->left : pCur->right;
        struct TreeNode *pNextQ = q->val < pCur->val ? pCur->left : pCur->right;
        
        if (pNextP == pNextQ)
            pCur = pNextP;
        else
            break;
    }
    
    return pRet;
}
